### PR TITLE
new regex for matching origins - issue #1981

### PR DESCRIPTION
**Pull request by zaach**

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
@zaach - I rescind my last comment.

File: lib/validate.js [hunk 1]
```diff
@@ -38,9 +38,24 @@ var types = {
...
+      [a-z0-9\-]{1,63}\b        // part will be 1-63 letters, numbers, or dashes
+        (?!\.$)                 // final part cannot end with a '.'
+        \.?                     // part followed by '.' unless final part
+    )+                          // one or more hostname parts
+    (:\d+)?                     // optional port
+    $/i;                        // end; case-insensitive
+    */
+    var regex = /^https?:\/\/(?=.{1,254}(?::|$))(?:(?!\d|-)(?![a-z0-9\-]{1,62}-(?:\.|:|$))[a-z0-9\-]{1,63}\b(?!\.$)\.?)+(:\d+)?$/i;
+    if (typeof x !== 'string' || !x.match(regex)) {
+      throw new Error("not a valid origin");
```
Comment (lloyd) [inline lib/validate.js:58]: OMG!  Monster regexp!  @fmarier, look what you've done!

File: lib/validate.js [hunk 2]
```diff
@@ -38,9 +38,24 @@ var types = {
...
+    (?:                         // match hostname part (<part>.<part>...)
+      (?!\d|-)                  // cannot start with a digit or dash
+      (?![a-z0-9\-]{1,62}-      // part cannot end with a dash
+        (?:\.|:|$))             // (end of part will be '.', ':', or end of str)
+      [a-z0-9\-]{1,63}\b        // part will be 1-63 letters, numbers, or dashes
+        (?!\.$)                 // final part cannot end with a '.'
+        \.?                     // part followed by '.' unless final part
+    )+                          // one or more hostname parts
+    (:\d+)?                     // optional port
+    $/i;                        // end; case-insensitive
```
Comment (lloyd) [inline lib/validate.js:54]: fantastic documentation of the RE.

**Comment by lloyd** [comment]
tests pass: http://travis-ci.org/#!/zaach/browserid/builds/1848454

r+

**Pull request merged**